skimage/restoration/inpaint.py

```diff
@@ -1,79 +1,213 @@
-    matrix_known = sparse.lil_matrix((np.sum(mask), out.size))
+def _get_neigh_coef(shape, center, coef_cache={}):
+    """Create biharmonic coefficients ndarray."""
+    key = (shape, center)
+    if key in coef_cache:
+        neigh_coef, coef_idx, coef_vals = coef_cache[key]
+    else:
+        # Create biharmonic coefficients ndarray
+        neigh_coef = np.zeros(shape)
+        neigh_coef[center] = 1
+        neigh_coef = laplace(laplace(neigh_coef))
+
+        # extract non-zero locations and values
+        coef_idx = np.where(neigh_coef)
+        coef_vals = neigh_coef[coef_idx]
+
+        # cache the result
+        coef_cache[key] = (neigh_coef, coef_idx, coef_vals)
+    return neigh_coef, coef_idx, coef_vals
+
 
-    # Find indexes of masked points in flatten array
-    mask_i = np.ravel_multi_index(np.where(mask), mask.shape)
+def _inpaint_biharmonic_single_region(image, mask, out, neigh_coef_full,
+                                      coef_vals, raveled_offsets, limits):
+    """Solve a (sparse) linear system corresponding to biharmonic inpainting.
 
-    # Find masked points and prepare them to be easily enumerate over
-    mask_pts = np.stack(np.where(mask), axis=-1)
+    This function creates a linear system of the form:
+
+    ``A @ u = b``
+
+    where ``A`` is a sparse matrix, ``b`` is a vector enforcing smoothness and
+    boundary constraints and ``u`` is the vector of inpainted values to be
+    (uniquely) determined by solving the linear system.
+
+    ``A`` is a sparse matrix of shape (n_mask, n_mask) where `n_mask``
+    corresponds to the number of non-zero values in ``mask`` (i.e. the number
+    of pixels to be inpainted). Each row in A will have a number of non-zero
+    values equal to the number of non-zero values in the biharmonic kernel,
+    ``neigh_coef_full``. In practice, biharmonic kernels with reduced extent
+    are used at the image borders. This matrix, ``A`` is the same for all
+    image channels (since the same inpainting mask is currently used for all
+    channels).
+
+    ``u`` is a dense matrix of shape ``(n_mask, n_channels)`` and represents
+    the vector of unknown values for each channel.
+
+    ``b`` is a dense matrix of shape ``(n_mask, n_channels)`` and represents
+    the desired output of convolving the solution with the biharmonic kernel.
+    At mask locations where there is no overlap with known values, ``b`` will
+    have a value of 0. This enforces the biharmonic smoothness constraint in
+    the interior of inpainting regions. For regions near the boundary that
+    overlap with known values, the entries in ``b`` enforce boundary conditions
+    designed to avoid discontinuity with the known values.
+    """
 
-    # Iterate over masked points
-    for mask_pt_n, mask_pt_idx in enumerate(mask_pts):
+    n_channels = out.shape[-1]
+    radius = neigh_coef_full.shape[0] // 2
+
+    edge_mask = np.ones(mask.shape, dtype=bool)
+    edge_mask[(slice(2, -2),) * mask.ndim] = 0
```

Comment:
Are the hard coded 2s here are intentional or is it supposed to be `radius`?

Reply:
Yes, it should be `radius`. (`radius` is always 2 for the biharmonic kernel used in the publication implemented here, though!).